docs/groovy/conceptual/positional-index-safety.md

```diff
@@ -0,0 +1,334 @@
+---
+title: Positional indices
+---
+
+When working with Deephaven tables, you may encounter positional indices (`i`, `ii`) and row keys (`k`) in formulas. Understanding when these are safe to use is crucial for avoiding incorrect results and runtime errors.
+
+## The fundamental rule
+
+**Positional indices (`i`, `ii`, `k`) and column array variables are unsafe on refreshing tables** and will cause `IllegalArgumentException` errors by default. The engine blocks these formulas to prevent incorrect results.
+
+- **`i`** and **`ii`** are positional indices that represent the current ordinal position of a row within a table.
+- **`k`** is the row key, a unique identifier for each row that may differ from positional index.
+- **Column array variables** allow direct access to column data arrays, which become invalid when table structure changes.
+
+## Key terms
+
+Before diving into the details, let's define the key concepts:
+
+- **Row key (`k`)**: A unique identifier assigned to each row in a table. Row keys are used internally by the engine to track rows and are typically sequential integers starting from 0.
+
+- **Positional index (`i`)**: The current position of a row within a table, starting from 0 for the first row. Unlike row keys, positional indices always reflect the current order of rows.
+
+- **Shifts**: When a table grows beyond its pre-allocated row key space, the engine may need to move (shift) other tables' rows to new row key positions to make room. The row data itself doesn't change, but the row keys do.
+
+- **Constituent tables**: The individual tables that are being merged together. Each constituent table gets its own allocated range of row keys in the merged result.
+
+- **Row key slots**: Pre-allocated ranges of row keys reserved for each constituent table. For example, table A might get slots 0-4095, while table B gets slots 4096-8191.
+
+- **Refreshing tables**: Tables that can change over time (live data). This includes tables created from live data sources (like `timeTable`, Kafka streams, database connections), tables that are the result of operations on refreshing tables, and any table where rows can be added, removed, or shifted after creation. In contrast, **static tables** never change after creation - they contain fixed data that remains constant.
+
+- **Append-only tables**: A special type of refreshing table where new rows are only added at the end (never inserted, moved, or deleted), making positional indices safe to use.
+
+## Why positional indices become unsafe
+
+When a refreshing table changes, several things can happen that make positional indices incorrect:
+
+### 1. Row additions
+
+When new rows are added to a table, the positional indices of existing rows may change. Consider this example:
+
+```groovy order=null
```

Comment:
I tested the equivalent Python code block and it worked just fine with the second part un-commented.